plugins/out_pgsql/pgsql.c

```diff
@@ -0,0 +1,270 @@
+                  ctx->db_hostname, PQerrorMessage(ctx->conn));
+
+        pgsql_conf_destroy(ctx);
+
+        return -1;
+    }
+
+    flb_info("[out_pgsql] host=%s port=%s dbname=%s OK",
+              ctx->db_hostname, ctx->db_port, ctx->db_name);
+    flb_output_set_context(ins, ctx);
+
+    temp = PQescapeIdentifier(ctx->conn, ctx->db_table, flb_sds_len(ctx->db_table));
+
+    if(temp == NULL) {
+        flb_error("[out_pgsql] failed to parse table name: %s", PQerrorMessage(ctx->conn));
+        pgsql_conf_destroy(ctx);
+        return -1;
+    } else {
+        flb_sds_destroy(ctx->db_table);
+        ctx->db_table = flb_sds_create(temp);
+        PQfreemem(temp);
+    }
+
+    flb_info("[out_pgsql] we check that the table %s exists, if not we create it", ctx->db_table);
+
+    str_len = 62 + flb_sds_len(ctx->db_table);
+
+    query = flb_malloc(str_len);
+    if(query == NULL) {
+        pgsql_conf_destroy(ctx);
+        return -1;
+    }
+
+    snprintf(query, str_len, "CREATE TABLE IF NOT EXISTS %s (data json);", ctx->db_table);
+    res = PQexec(ctx->conn, query);
+    flb_free(query);
+
+    if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+        flb_error("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+        pgsql_conf_destroy(ctx);
+        return -1;
+    }
+
+    PQclear(res);
+
+    flb_info("[out_pgsql] switching postgresql connection to non-blocking mode");
+    if(PQsetnonblocking(ctx->conn, 1) != 0) {
+        flb_error("[out_pgsql] non-blocking mode not set, not good");
+        pgsql_conf_destroy(ctx);
+        return -1;
+    }
+
+    return 0;
+}
+
+static void cb_pgsql_flush(const void *data, size_t bytes,
+                            const char *tag, int tag_len,
+                            struct flb_input_instance *i_ins,
+                            void *out_context,
+                            struct flb_config *config)
+{
+    struct flb_pgsql_config *ctx = out_context;
+    flb_sds_t json;
+    PGresult *res = NULL;
+    char *query = NULL;
+    size_t str_len;
+
+    if(PQconsumeInput(ctx->conn) == 0 && PQisBusy(ctx->conn) == 1) {
+        flb_debug("[out_pgsql] Some command may still be running");
+    }
+
+    /*
+      PQreset()
+      This function will close the connection to the server and attempt to reestablish a new connection
+      to the same server, using all the same parameters previously used. This might be useful for error
+      recovery if a working connection is lost.
+     */
+    if(PQstatus(ctx->conn) != CONNECTION_OK) {
+        PQreset(ctx->conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+
+    json = flb_pack_msgpack_to_json_format(data, bytes,
+                                           FLB_PACK_JSON_FORMAT_JSON,
+                                           FLB_PACK_JSON_DATE_DOUBLE,
+                                           ctx->timestamp_key);
+    if(json == NULL) {
+        FLB_OUTPUT_RETURN(FLB_RETRY);
```

Comment:
nit: might be worth adding an error message

Reply:
hooo yeah! done! :D fixed!